### ISSUE TITLE
`useContract`

### ISSUE THREAD

**Issue by Dhaiwat10** [Developer-DAO/web3-ui#39]

**Comment by swellander** [comment]
Hey @Dhaiwat10, can I be assigned to this one? Working on a rough draft now. Thanks!

**Comment by crondinini** [comment]
Hey @swellander, are you working on both this one and https://github.com/Developer-DAO/web3-ui/issues/17? I want to make sure both are in progress.

**Comment by crondinini** [comment]
Copy/pasting relevant comment from another issue: 
https://github.com/Developer-DAO/web3-ui/issues/36#issuecomment-983508910

> Couldn't we use TypeChain to generate the type definitions?

I wanted to +1 on this.

**Comment by swellander** [comment]
@crondinini yep currently working on this one! It looked like @RoguePenguins was working on [#17](https://github.com/Developer-DAO/web3-ui/issues/17) so I switched to this issue.

As for TypeChain, could you maybe elaborate on a high level how the integration would work? I've never used the library, but after looking at it a bit I'm still unsure how we would generate types dynamically based off of the `abi` the user passes to `useContract(address, abi)`. Thanks!

**Comment by Dhaiwat10** [comment]
@swellander have you started working on this? Are you blocked in any way? How can I help you?

**Comment by swellander** [comment]
Hi @Dhaiwat10, thanks for offering to help. I'll share a bit of what I've been thinking so far and maybe you can let me know if I'm on the right track.

### Context
The current version of the `useContract` hook that I have right now looks like this:
```ts
function useContract(address: string, abi) {
    const context = React.useContext(Web3Context);
    const [contract, setContract] = React.useState({});
    React.useEffect(() => {
        if (context) {
          const newContract = new Contract(
            address,
            abi,
            context.signer 
          )
          const contractInterface = Object.values(newContract.interface.functions).reduce((accumulator, funcFragment) => {
            return {...accumulator, [funcFragment.name]: newContract[funcFragment.name]}
          }, {})
          setContract(contractInterface);
        }
    }, [context])
    return contract;
}
```
The hook returns `contractInterface` which is intended to be an object with all of the contract's available methods, accessed through an [`ethers.js` `Contract` instance](https://docs.ethers.io/v5/api/contract/contract/#Contract).

### Blocking Questions
1. I'm confused about the `readContract` and `writeContract` methods you mentioned in [this issue's description](https://github.com/Developer-DAO/web3-ui/issues/39#issue-1066421826). Are these like proxy methods that would accept the name of a property or method on the contract?
2. I'm struggling to come up with a way to type the contract passed to `useContract()`. The [Typechain](https://github.com/dethcrypto/TypeChain) library mentioned by @christiananese and @crondinini seems like a promising solution, but I haven't been able to think of an effective way to integrate it (see [my previous comment on this issue](https://github.com/Developer-DAO/web3-ui/issues/39#issuecomment-986101895)).

Thanks again for your help!

Also, let me know if you think it would be more appropriate to have this discussion in a draft PR instead.

**Comment by Dhaiwat10** [comment]
@swellander that's a great starting point, thanks for writing this up. Re. your questions:

1. > Are these like proxy methods that would accept the name of a property or method on the contract?

Yes. But now that I think about it, it's better to just do something like this for now:

```tsx
const contract = useContract({ address: '0xxxxxababa', abi: '[]' })

// read
contract.greeting();

// write
contract.setGreeting('New greeting');

// payable
contract.transferTo('0x0x0x0x0',  { value: '1000000000000000000' }) // value amount in wei
```

Basically the same as calling the contract using `ethers`. I don't think making a distinction between `read` and `write` is worth it right now.

2. Gotcha. How about you create a draft PR and we continue the discussion there? Type definitions are not a priority right now, FWIW. We can add them through a separate PR as they will also change the hooks' interface. :)

**Comment by swellander** [comment]
@Dhaiwat10 got it. Thanks for the reply!
